tests/dialects/test_builtin.py

```diff
@@ -45,3 +49,90 @@ def test_DenseArrayBase_verifier_failure():
         DenseArrayBase([i32, ArrayAttr.from_list([FloatData(0.0)])])
     assert err.value.args[0] == ("dense array of integer element type "
                                  "should only contain integers")
+
+
+def test_vector_rank_constraint_verify():
+    vector_type = VectorType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorRankConstraint(2)
+
+    constraint.verify(vector_type)
+
+
+def test_vector_rank_constraint_rank_mismatch():
+    vector_type = VectorType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorRankConstraint(3)
+
+    with pytest.raises(VerifyException) as e:
+        constraint.verify(vector_type)
+    assert e.value.args[0] == "Expected vector rank to be 3, got 2."
+
+
+def test_vector_rank_constraint_attr_mismatch():
+    memref_type = MemRefType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorRankConstraint(3)
+
+    with pytest.raises(VerifyException) as e:
+        constraint.verify(memref_type)
+    assert e.value.args[
+        0] == "!memref<[1 : !index, 2 : !index], !i32> should be of type VectorType."
+
+
+def test_vector_base_type_constraint_verify():
+    vector_type = VectorType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorBaseTypeConstraint(i32)
+
+    constraint.verify(vector_type)
+
+
+def test_vector_base_type_constraint_type_mismatch():
+    vector_type = VectorType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorBaseTypeConstraint(i64)
+
+    with pytest.raises(VerifyException) as e:
+        constraint.verify(vector_type)
+    assert e.value.args[0] == "Expected vector type to be !i64, got !i32."
+
+
+def test_vector_base_type_constraint_attr_mismatch():
+    memref_type = MemRefType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorBaseTypeConstraint(i32)
+
+    with pytest.raises(VerifyException) as e:
+        constraint.verify(memref_type)
+    assert e.value.args[
+        0] == "!memref<[1 : !index, 2 : !index], !i32> should be of type VectorType."
+
+
+def test_vector_base_type_and_rank_constraint_verify():
+    vector_type = VectorType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorBaseTypeAndRankConstraint(i32, 2)
+
+    constraint.verify(vector_type)
+
+
+def test_vector_base_type_and_rank_constraint_base_type_mismatch():
+    vector_type = VectorType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorBaseTypeAndRankConstraint(i64, 2)
+
+    with pytest.raises(VerifyException) as e:
+        constraint.verify(vector_type)
+    assert e.value.args[0] == "Expected vector type to be !i64, got !i32."
+
+
+def test_vector_base_type_and_rank_constraint_rank_mismatch():
+    vector_type = VectorType.from_element_type_and_shape(i32, [1, 2])
+    constraint = VectorBaseTypeAndRankConstraint(i32, 3)
```

Comment:
If both are wrong, you just report the first error, right? Not sure, maybe it would be cool to report both? But I guess this would involve weird checking (first both, then individually). What do you think?

Reply:
Yes, I don't see a straightforward way to do this atm. If we do decide to move ahead by implementing this with `AllOf` [which I think is a good idea], we'll have to make some changes in its implementation as well to get this done.

Reply:
This should be fixed now!